xla/service/gpu/gpu_all_gather_optimizer.cc

```diff
@@ -0,0 +1,194 @@
+                                       ? user->mutable_operand(1)
+                                       : user->mutable_operand(0);
+  if (allgatherRight->opcode() != HloOpcode::kAllGather) {
+    VLOG(2) << "Binary op's right operand is not all-gather "
+            << allgatherRight->ToString();
+    return std::nullopt;
+  }
+  // right side all-gather is also subject to removal
+  // and should not contain more than 1 users
+  if (allgatherRight->user_count() != 1) {
+    VLOG(2) << "right side all-gather user_count > 1 "
+            << allgatherRight->ToString();
+    return std::nullopt;
+  }
+  HloInstruction* rightReduceScatter = allgatherRight->mutable_operand(0);
+  
+  // we need to traverse the right branch backwards in search of 
+  // a reduce scatter collective
+  while (rightReduceScatter->opcode() != HloOpcode::kReduceScatter &&
+             rightReduceScatter->operand_count() > 0 ) {
+        rightReduceScatter = rightReduceScatter->mutable_operand(0);
+      }
+
+  if (rightReduceScatter->opcode() != HloOpcode::kReduceScatter) {
+    VLOG(2)
+        << "Binary op's right operand path doesn not include reduce scatter "
+        << rightReduceScatter->ToString();
+    return std::nullopt;
+  }
+  if (!ReplicaGroupsEqual(rs->replica_groups(),
+                          rightReduceScatter->replica_groups())) {
+    VLOG(2)
+        << "Reduce-Scatters in two branches don't have similar replica groups"
+        << rightReduceScatter->ToString();
+    return std::nullopt;
+  }
+  AllGatherReduceScatterSpec spec;
+  spec.all_gather_left = allgatherLeft;
+  spec.all_gather_right = allgatherRight;
+  spec.binary_op = binaryOp;
+  spec.right_source = allgatherRight->mutable_operand(0);
+
+  return spec;
+}
+
+StatusOr<bool> AllGatherOptimizer::Run(
+    HloModule* module,
+    const absl::flat_hash_set<absl::string_view>& execution_threads) {
+  const HloModuleConfig& config = module->config();
+  int64_t next_channel_id = hlo_query::NextChannelId(*module);
+
+  bool changed = false;
+  for (HloComputation* computation :
+       module->MakeNonfusionComputations(execution_threads)) {
+    for (HloInstruction* instruction :
+         computation->MakeInstructionPostOrder()) {
+      if (instruction->opcode() != HloOpcode::kReduceScatter ||
+          !instruction->shape().IsArray()) {
+        continue;
+      }
+      auto* rs = Cast<HloReduceScatterInstruction>(instruction);
+      // get the graph structure for which the all-gather ops
+      // are combined into a single call after the binary op
+      auto rs_spec = MatchReduceScatter(rs);
+
+      if (!rs_spec) {
+        VLOG(2) << "Cannot match all-gather combining optimziation "
+                << rs->ToString();
+        continue;
+      }
+      auto index_in_full_shape =
+          computation->AddInstruction(HloInstruction::CreateBinary(
+              rs->shape(), rs_spec->binary_op->opcode(), rs,
```

Comment:
I think it would be clearer to see what is happening if you change this to:

```
auto index_in_full_shape =
          computation->AddInstruction(HloInstruction::CreateBinary(
              rs_spec->all_gather_left->operand(0)->shape(), rs_spec->binary_op->opcode(), rs_spec->all_gather_left->operand(0), 
              rs_spec->all_gather_right->operand(0)));
```
Because as far as I can tell, nothing here depends on the ReduceScatter op, what matters are the operands of the two all-gathers.

Reply:
Great catch! I applied your suggestion resulting in dramatic simplification!